src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs

```diff
@@ -1084,144 +1185,290 @@ async Task StartProcessingInvocationMessages(ChannelReader<InvocationMessage> in
+                    try
+                    {
+                        _state.ChangeState(HubConnectionState.Reconnecting, HubConnectionState.Disconnected);
+
+                        CompleteClose(GetOperationCanceledException("Connection stopped during reconnect delay. Done reconnecting.", ex, _state.StopCts.Token));
+                    }
+                    finally
+                    {
+                        _state.ReleaseConnectionLock();
+                    }
+
+                    return;
                 }
+
+                await _state.WaitConnectionLockAsync();
+                try
+                {
+                    SafeAssert(ReferenceEquals(_state.CurrentConnectionStateUnsynchronized, null),
+                        "Someone other than Reconnect set the connection state!");
+
+                    // HandshakeAsync already checks ReconnectingConnectionState.StopCts.Token.
+                    await StartAsyncCore(CancellationToken.None);
+
+                    Log.Reconnected(_logger, previousReconnectAttempts, DateTime.UtcNow - reconnectStartTime);
+
+                    _state.ChangeState(HubConnectionState.Reconnecting, HubConnectionState.Connected);
+
+                    RunReconnectedEvent();
+                    return;
+                }
+                catch (Exception ex)
+                {
+                    retryReason = ex;
+
+                    Log.ReconnectAttemptFailed(_logger, ex);
+
+                    if (_state.StopCts.IsCancellationRequested)
+                    {
+                        Log.ReconnectingStoppedDuringReconnectAttempt(_logger);
+
+                        _state.ChangeState(HubConnectionState.Reconnecting, HubConnectionState.Disconnected);
+
+                        CompleteClose(GetOperationCanceledException("Connection stopped during reconnect attempt. Done reconnecting.", ex, _state.StopCts.Token));
+                        return;
+                    }
+                }
+                finally
+                {
+                    _state.ReleaseConnectionLock();
+                }
+
+                nextRetryDelay = GetNextRetryDelay(previousReconnectAttempts++, DateTime.UtcNow - reconnectStartTime, retryReason);
+            }
+
+            await _state.WaitConnectionLockAsync();
+            try
+            {
+                SafeAssert(ReferenceEquals(_state.CurrentConnectionStateUnsynchronized, null),
+                    "Someone other than Reconnect set the connection state!");
+
+                var elapsedTime = DateTime.UtcNow - reconnectStartTime;
+                Log.ReconnectAttempsExhausted(_logger, previousReconnectAttempts, elapsedTime);
+
+                _state.ChangeState(HubConnectionState.Reconnecting, HubConnectionState.Disconnected);
+
+                var message = $"Reconnect retries have been exhausted after {previousReconnectAttempts} failed attempts and {elapsedTime} elapsed. Disconnecting.";
```

Comment:
I think "00:01:22 elapsed" is pretty understandable. I trust the default settings for TimeSpan.ToString() to produce something reasonable. Since I don't see an alternative format suggested, I'll leave this open to future PRs.